File: backend-csharp/Services/SessionService.cs
using Microsoft.EntityFrameworkCore;
using Picture2Text.Api.Data;
using Picture2Text.Api.Models;

namespace Picture2Text.Api.Services;

/// <summary>
/// 使用者會話管理服務
/// </summary>
public class SessionService
{
    private readonly ApplicationDbContext _context;
    private readonly IConfiguration _configuration;
    private readonly int _sessionExpirationDays;

    public SessionService(ApplicationDbContext context, IConfiguration configuration)
    {
        _context = context;
        _configuration = configuration;
        _sessionExpirationDays = int.Parse(configuration["Jwt:RefreshTokenExpirationDays"] ?? "7");
    }

    /// <summary>
    /// 創建新會話
    /// </summary>
    public async Task<UserSession> CreateSessionAsync(
        int userId,
        int? refreshTokenId,
        string? ipAddress,
        string? userAgent,
        string? deviceName = null)
    {
        var session = new UserSession
        {
            UserId = userId,
            SessionId = Guid.NewGuid().ToString(),
            RefreshTokenId = refreshTokenId,
            DeviceName = deviceName ?? ParseDeviceName(userAgent),
            IpAddress = ipAddress,
            UserAgent = userAgent,
            LoginAt = DateTime.UtcNow,
            LastActivityAt = DateTime.UtcNow,
            ExpiresAt = DateTime.UtcNow.AddDays(_sessionExpirationDays),
            IsActive = true
        };

        _context.UserSessions.Add(session);
        await _context.SaveChangesAsync();

        return session;
    }

    /// <summary>
    /// 更新會話活躍時間
    /// </summary>
    public async Task UpdateSessionActivityAsync(string sessionId)
    {
        var session = await _context.UserSessions
            .FirstOrDefaultAsync(s => s.SessionId == sessionId && s.IsActive);

        if (session != null)
        {
            session.LastActivityAt = DateTime.UtcNow;
            await _context.SaveChangesAsync();
        }
    }

    /// <summary>
    /// 結束會話（登出）
    /// </summary>
    public async Task<bool> EndSessionAsync(string sessionId)
    {
        var session = await _context.UserSessions
            .FirstOrDefaultAsync(s => s.SessionId == sessionId && s.IsActive);

        if (session == null)
        {
            return false;
        }

        session.IsActive = false;
        session.LogoutAt = DateTime.UtcNow;
        await _context.SaveChangesAsync();

        return true;
    }

    /// <summary>
    /// 結束使用者的所有會話
    /// </summary>
    public async Task EndAllUserSessionsAsync(int userId)
    {
        var sessions = await _context.UserSessions
            .Where(s => s.UserId == userId && s.IsActive)
            .ToListAsync();

        foreach (var session in sessions)
        {
            session.IsActive = false;
            session.LogoutAt = DateTime.UtcNow;
        }

        await _context.SaveChangesAsync();
    }

    /// <summary>
    /// 取得使用者的所有活躍會話
    /// </summary>
    public async Task<List<UserSession>> GetUserActiveSessionsAsync(int userId)
    {
        return await _context.UserSessions
            .Where(s => s.UserId == userId && s.IsActive && s.ExpiresAt > DateTime.UtcNow)
            .OrderByDescending(s => s.LastActivityAt)
            .ToListAsync();
    }

    /// <summary>
    /// 清理過期會話
    /// </summary>
    public async Task CleanupExpiredSessionsAsync()
    {
        var expiredSessions = await _context.UserSessions
            .Where(s => s.IsActive && s.ExpiresAt < DateTime.UtcNow)
            .ToListAsync();

        foreach (var session in expiredSessions)
        {
            session.IsActive = false;
            session.LogoutAt = DateTime.UtcNow;
        }

        await _context.SaveChangesAsync();
    }

    /// <summary>
    /// 從 User-Agent 解析裝置名稱
    /// </summary>
    private static string ParseDeviceName(string? userAgent)
    {
        if (string.IsNullOrEmpty(userAgent))
        {
            return "未知裝置";
        }

        // 簡單的裝置識別邏輯
        if (userAgent.Contains("iPhone", StringComparison.OrdinalIgnoreCase))
            return "iPhone";
        if (userAgent.Contains("iPad", StringComparison.OrdinalIgnoreCase))
            return "iPad";
        if (userAgent.Contains("Android", StringComparison.OrdinalIgnoreCase))
            return "Android 裝置";
        if (userAgent.Contains("Windows", StringComparison.OrdinalIgnoreCase))
            return "Windows 電腦";
        if (userAgent.Contains("Macintosh", StringComparison.OrdinalIgnoreCase))
            return "Mac 電腦";
        if (userAgent.Contains("Linux", StringComparison.OrdinalIgnoreCase))
            return "Linux 電腦";

        return "未知裝置";
    }
}

File: backend-nestjs/src/services/session.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, LessThan } from 'typeorm';
import { ConfigService } from '@nestjs/config';
import { v4 as uuidv4 } from 'uuid';
import { UserSession } from '../entities';
import { SessionInfo } from '../dto/response';

@Injectable()
export class SessionService {
  private readonly logger = new Logger(SessionService.name);
  private readonly refreshTokenExpirationDays: number;

  constructor(
    @InjectRepository(UserSession)
    private readonly sessionRepository: Repository<UserSession>,
    private readonly configService: ConfigService,
  ) {
    this.refreshTokenExpirationDays = this.configService.get<number>('jwt.refreshTokenExpirationDays')!;
  }

  /**
   * 建立新 Session
   */
  async createSessionAsync(
    userId: number,
    refreshTokenId: number,
    ipAddress?: string,
    userAgent?: string,
  ): Promise<UserSession> {
    const sessionId = uuidv4();
    const now = new Date();
    const expiresAt = new Date();
    expiresAt.setDate(expiresAt.getDate() + this.refreshTokenExpirationDays);

    const session = this.sessionRepository.create({
      UserId: userId,
      SessionId: sessionId,
      RefreshTokenId: refreshTokenId,
      DeviceName: this.parseDeviceName(userAgent),
      IpAddress: ipAddress,
      UserAgent: userAgent,
      LoginAt: now,
      LastActivityAt: now,
      ExpiresAt: expiresAt,
      IsActive: true,
    });

    await this.sessionRepository.save(session);

    this.logger.log(`Created session ${sessionId} for user ${userId}`);

    return session;
  }

  /**
   * 更新 Session 活動時間
   */
  async updateSessionActivityAsync(sessionId: string): Promise<void> {
    await this.sessionRepository.update(
      { SessionId: sessionId, IsActive: true },
      { LastActivityAt: new Date() },
    );
  }

  /**
   * 結束 Session
   */
  async endSessionAsync(sessionId: string, userId: number): Promise<boolean> {
    const result = await this.sessionRepository.update(
      { SessionId: sessionId, UserId: userId, IsActive: true },
      { IsActive: false, LogoutAt: new Date() },
    );

    if (result.affected && result.affected > 0) {
      this.logger.log(`Ended session ${sessionId} for user ${userId}`);
      return true;
    }

    return false;
  }

  /**
   * 結束使用者的所有 Session
   */
  async endAllUserSessionsAsync(userId: number): Promise<number> {
    const result = await this.sessionRepository.update(
      { UserId: userId, IsActive: true },
      { IsActive: false, LogoutAt: new Date() },
    );

    const count = result.affected || 0;
    this.logger.log(`Ended ${count} sessions for user ${userId}`);

    return count;
  }

  /**
   * 取得使用者的活躍 Session 列表
   */
  async getUserActiveSessionsAsync(userId: number, currentSessionId?: string): Promise<SessionInfo[]> {
    const sessions = await this.sessionRepository.find({
      where: { UserId: userId, IsActive: true },
      order: { LastActivityAt: 'DESC' },
    });

    return sessions.map((session) => ({
      SessionId: session.SessionId,
      DeviceName: session.DeviceName || undefined,
      IpAddress: session.IpAddress || undefined,
      LoginAt: session.LoginAt,
      LastActivityAt: session.LastActivityAt,
      ExpiresAt: session.ExpiresAt,
      IsCurrent: session.SessionId === currentSessionId,
    }));
  }

  /**
   * 根據 Session ID 取得 Session
   */
  async getSessionByIdAsync(sessionId: string): Promise<UserSession | null> {
    return this.sessionRepository.findOne({
      where: { SessionId: sessionId },
      relations: ['User'],
    });
  }

  /**
   * 更新 Session 的 Refresh Token ID
   */
  async updateSessionRefreshTokenAsync(sessionId: string, refreshTokenId: number): Promise<void> {
    const expiresAt = new Date();
    expiresAt.setDate(expiresAt.getDate() + this.refreshTokenExpirationDays);

    await this.sessionRepository.update(
      { SessionId: sessionId },
      {
        RefreshTokenId: refreshTokenId,
        LastActivityAt: new Date(),
        ExpiresAt: expiresAt,
      },
    );
  }

  /**
   * 清理過期的 Session
   */
  async cleanupExpiredSessionsAsync(): Promise<number> {
    const result = await this.sessionRepository.update(
      { IsActive: true, ExpiresAt: LessThan(new Date()) },
      { IsActive: false },
    );

    const count = result.affected || 0;
    if (count > 0) {
      this.logger.log(`Marked ${count} expired sessions as inactive`);
    }

    return count;
  }

  /**
   * 解析裝置名稱
   */
  private parseDeviceName(userAgent?: string): string {
    if (!userAgent) {
      return 'Unknown Device';
    }

    const ua = userAgent.toLowerCase();

    // 偵測瀏覽器
    let browser = 'Unknown Browser';
    if (ua.includes('edg/')) {
      browser = 'Edge';
    } else if (ua.includes('chrome/')) {
      browser = 'Chrome';
    } else if (ua.includes('firefox/')) {
      browser = 'Firefox';
    } else if (ua.includes('safari/') && !ua.includes('chrome')) {
      browser = 'Safari';
    } else if (ua.includes('opera') || ua.includes('opr/')) {
      browser = 'Opera';
    }

    // 偵測作業系統
    let os = 'Unknown OS';
    if (ua.includes('windows')) {
      os = 'Windows';
    } else if (ua.includes('macintosh') || ua.includes('mac os')) {
      os = 'macOS';
    } else if (ua.includes('linux') && !ua.includes('android')) {
      os = 'Linux';
    } else if (ua.includes('android')) {
      os = 'Android';
    } else if (ua.includes('iphone') || ua.includes('ipad')) {
      os = 'iOS';
    }

    return `${browser} on ${os}`;
  }
}
